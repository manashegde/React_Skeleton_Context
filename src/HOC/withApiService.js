import React from "react";

const xbcGatewayServicesElem = document.getElementById("xbcGatewayServicesEp");
const xbcGatewayUrl = xbcGatewayServicesElem
  ? xbcGatewayServicesElem.getAttribute("di")
  : "/nic-xbc-gateway/xbc/gateway/";

const nicIdentityServiceElem = document.getElementById("nicIdentityService");
const nicIdentityServiceUrl = nicIdentityServiceElem
  ? nicIdentityServiceElem.getAttribute("di")
  : "http://10.0.1.6:50010/mocks-gateway/xbc/mocks/gateway/";

export const urls = {
  login: `${nicIdentityServiceUrl}login/v1`,
  refreshToken: `${nicIdentityServiceUrl}token/refresh/v1`,
  postLogin: `${xbcGatewayUrl}security/post-login`,

  crossing: {
    arrival: {
      search: {
        citizen: `${xbcGatewayUrl}crossing/search/arrival/citizen`,
        alien: `${xbcGatewayUrl}crossing/search/arrival/alien`
      },
      complete: {
        citizen: `${xbcGatewayUrl}crossing/arrival/complete/citizen`,
        alien: `${xbcGatewayUrl}crossing/arrival/complete/alien`
      }
    },
    departure: {
      search: {
        citizen: `${xbcGatewayUrl}crossing/search/departure/citizen`
      },
      complete: {
        citizen: `${xbcGatewayUrl}crossing/departure/complete/citizen`
      }
    },
    specialSearch: `${xbcGatewayUrl}crossing/search/special`,
    lookup: `${xbcGatewayUrl}lookup/ref-data/`, //Bcr269BusRule/readAll`,
    fetchTripInfo: `${xbcGatewayUrl}crossing/search/tripInfo/fetch`,
    updateTripInfo: `${xbcGatewayUrl}crossing/search/tripInfo/update`,
    searchCwl: `${xbcGatewayUrl}crossing/cwl/search`,
    updateCwl: `${xbcGatewayUrl}crossing/cwl/update`,
    searchCwlDetails: `${xbcGatewayUrl}crossing/cwl/details`
  }
};

export function makeLookupUrl({ entity, pk, isReadAll }) {
  let out = urls.crossing.lookup + entity;
  if (isReadAll) {
    return out + "/readAll";
  } else {
    return out + "/read" + (pk ? `/${pk}` : "");
  }
}

export function makeCrossingUrlFor(action, travelerType, crossingType) {
  try {
    const { crossing } = urls;
    const result = [crossingType, action, travelerType].reduce((prev, curr) => {
      return prev[curr];
    }, crossing);
    console.log("------crossing url:", result);
    return result;
  } catch (e) {
    throw new Error(
      "failed to make crossing url: missing action, travelerType, or crossingType"
    );
  }
}

export function parseJwt(token) {
  var base64Url = token.split(".")[1];
  var base64 = base64Url.replace(/-/g, "+").replace(/_/g, "/");
  return JSON.parse(window.atob(base64));
}

export function post({ url, data, method }) {
  const jwt2 = sessionStorage.getItem("jwt2") || "";
  let headers = {
    "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8"
  };
  if (jwt2) headers["Authorization"] = `Bearer ${jwt2}`;
  return fetch(url, {
    body: JSON.stringify(data), // must match 'Content-Type' header
    headers,
    method: method || "POST", // *GET, POST, PUT, DELETE, etc.
    referrer: "no-referrer" // *client, no-referrer
  })
    .then(res => res.json().catch(err => res))
    .then(res => {
      if (res.status && res.status !== 200) {
        throw res;
      } else {
        return res;
      }
    })
    .catch(e => {
      throw e.message || e.statusText || "unknown post error";
    });
}

export function get(url) {
  return fetch(url)
    .then(res => res.json())
    .catch(e => {
      console.error("--------api get:", e);
      throw e.message || e.statusText || "unknown get error";
    });
}

//DO NOT USE THIS ANYMORE.
//USE POST AND GET METHODS FROM SESSIONCONTEXT INSTEAD
export default function withApiService(WrappedComponent) {
  class C extends React.Component {
    render() {
      return (
        <WrappedComponent
          {...this.props}
          apiService={{ post, get, urls, parseJwt }}
        />
      );
    }
  }
  C.displayName = `withApiService(${WrappedComponent.displayName ||
    WrappedComponent.name})`;
  return C;
}
