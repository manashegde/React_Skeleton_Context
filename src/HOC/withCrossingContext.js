import React from "react";
import { CrossingContextConsumer } from "contexts/crossing";

export default function withCrossingContext(WrappedComponent) {
  class C extends React.Component {
    render() {
      return (
        <CrossingContextConsumer>
          {crossingContext => (
            <WrappedComponent
              {...this.props}
              crossingContext={crossingContext}
            />
          )}
        </CrossingContextConsumer>
      );
    }
  }
  C.displayName = `withCrossingContext(${WrappedComponent.displayName ||
    WrappedComponent.name})`;
  return C;
}
