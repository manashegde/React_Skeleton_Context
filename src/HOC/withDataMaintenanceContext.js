import React from "react";
import { DataMaintContextConsumer } from "contexts/dataMaintenance";

export default function withDataMaintenanceContext(WrappedComponent) {
  class C extends React.Component {
    render() {
      return (
        <DataMaintContextConsumer>
          {dataMaintenanceContext => (
            <WrappedComponent
              {...this.props}
              dataMaintenanceContext={dataMaintenanceContext}
            />
          )}
        </DataMaintContextConsumer>
      );
    }
  }
  C.displayName = `withDataMaintenanceContext(${WrappedComponent.displayName ||
    WrappedComponent.name})`;
  return C;
}
