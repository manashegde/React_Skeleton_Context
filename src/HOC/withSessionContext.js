import React from "react";
import { SessionContextConsumer } from "contexts/session";

export default function withSessionContext(WrappedComponent) {
  class C extends React.Component {
    render() {
      return (
        <SessionContextConsumer>
          {session => (
            <WrappedComponent {...this.props} sessionContext={session} />
          )}
        </SessionContextConsumer>
      );
    }
  }
  C.displayName = `withSessionContext(${WrappedComponent.displayName ||
    WrappedComponent.name})`;
  return C;
}
