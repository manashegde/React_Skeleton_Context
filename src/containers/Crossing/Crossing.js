import React from "react";
import { Route, Switch } from "react-router-dom";
import Loadable from "react-loadable";
import NotFound from "components/NotFound";
import Loading from "components/Loading";
import ErrorBoundary from "components/ErrorBoundary";
import { routePaths } from "containers/Routes";
import "./Crossing.css";

const AsyncSearch = Loadable({
  loader: () => import("containers/Crossing/Search"),
  loading: Loading
});

const AsyncSpecialSearch = Loadable({
  loader: () => import("containers/Crossing/Search/SpecialSearch"),
  loading: Loading
});

const AsyncSamePassport = Loadable({
  loader: () => import("containers/Crossing/Search/SamePassport"),
  loading: Loading
});

const AsyncSearchResult = Loadable({
  loader: () => import("containers/Crossing/SearchResult"),
  loading: Loading
});

const AsyncTripInfo = Loadable({
  loader: () => import("containers/Crossing/TripInfo"),
  loading: Loading
});

const AsyncHitReport = Loadable({
  loader: () => import("containers/Crossing/HitReport"),
  loading: Loading
});

const AsyncMiscreantDetail = Loadable({
  loader: () => import("containers/Crossing/HitReport/MiscreantDetail"),
  loading: Loading
});

const AsyncWantedDetail = Loadable({
  loader: () => import("containers/Crossing/HitReport/WantedDetail"),
  loading: Loading
});

const AsyncAfisDetail = Loadable({
  loader: () => import("containers/Crossing/HitReport/AfisDetail"),
  loading: Loading
});

const AsyncAfisDetailEnrollment = Loadable({
  loader: () =>
    import("containers/Crossing/HitReport/AfisDetail/AfisDetailEnrollment"),
  loading: Loading
});

const AsyncFingerVerification = Loadable({
  loader: () => import("containers/Crossing/FingerVerification"),
  loading: Loading
});

class Crossing extends React.Component {
  render() {
    return (
      <ErrorBoundary>
        <div id="crossing">
          <Switch>
            <Route
              exact
              path={routePaths.crossing.search}
              component={AsyncSearch}
            />
            <Route
              exact
              path={routePaths.crossing.specialSearch}
              component={AsyncSpecialSearch}
            />
            <Route
              exact
              path={routePaths.crossing.samePassport}
              component={AsyncSamePassport}
            />
            <Route
              exact
              path={routePaths.crossing.tripInfo}
              component={AsyncTripInfo}
            />
            <Route
              exact
              path={routePaths.crossing.searchResult}
              component={AsyncSearchResult}
            />
            <Route
              exact
              path={routePaths.crossing.fingerVerification}
              component={AsyncFingerVerification}
            />
            <Route
              exact
              path={routePaths.crossing.hitReport}
              component={AsyncHitReport}
            />
            <Route
              exact
              path={routePaths.crossing.miscreantDetail}
              component={AsyncMiscreantDetail}
            />
            <Route
              exact
              path={routePaths.crossing.wantedDetail}
              component={AsyncWantedDetail}
            />
            <Route
              exact
              path={routePaths.crossing.afisDetail}
              component={AsyncAfisDetail}
            />
            <Route
              exact
              path={routePaths.crossing.afisDetailEnrollment}
              component={AsyncAfisDetailEnrollment}
            />
            <Route component={NotFound} />
          </Switch>
        </div>
      </ErrorBoundary>
    );
  }
}

export default Crossing;
