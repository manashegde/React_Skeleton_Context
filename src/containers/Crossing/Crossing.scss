@import "styles/colors.scss";

#crossing {
  // display: flex;
  // flex-direction: column;
  // align-items: center;
  height: 100%;
  overflow: auto;
  .label,
  .subtitle {
    // color: $white;
  }
  // padding: 2rem;
  // .form-box {
  //   background: $gray-gradient;
  //   padding: 2rem 4rem 4rem 4rem;
  //   .table thead th {
  //     color: #7a7a7a;
  //   }
  //   .select,
  //   select {
  //     width: 100%;
  //   }
  //   .subtitle {
  //     color: $white;
  //     text-align: center;
  //   }
  //   label {
  //     font-weight: 400;
  //     color: $white;
  //   }
  //   .table {
  //     thead th {
  //       background-color: rgba(85, 111, 127, 0.95);
  //       border: none;
  //       color: $white;
  //       position: sticky;
  //       top: -1px;
  //     }
  //   }
  // }
}
@media only screen and (max-width: 565px) {
  #crossing {
    align-items: stretch;
  }
}
