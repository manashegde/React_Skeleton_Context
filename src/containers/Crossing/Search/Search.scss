@import "styles/colors.scss";

#search {
  input.input,
  button[name="travelerType"] {
    border-radius: 1px;
  }
  padding: 0.75rem;
  // height: 100%;
  // width: 100%;

  // display: flex;
  // flex-direction: column;
  // justify-content: flex-start;
  // align-items: center;
  // padding-top: 3rem;
  // padding-bottom: 2rem;
  // overflow: auto;

  // .form-box {
  //   background: $gray-gradient;
  //   .subtitle {
  //     color: $white;
  //     text-align: center;
  //   }
  //   label {
  //     font-weight: 400;
  //     color: $white;
  //   }
  // }
}
