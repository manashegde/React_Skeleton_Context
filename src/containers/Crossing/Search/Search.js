import React from "react";
import moment from "moment";
import {
  travelerTypes,
  crossingTypes,
  crossingActions
} from "contexts/crossing";
import Link from "components/Link";
import { setInputValue } from "utils/helpers";
import compose from "utils/compose";
import withSessionContext from "HOC/withSessionContext";
import withCrossingContext from "HOC/withCrossingContext";
// import withApiService, { makeCrossingUrlFor } from "HOC/withApiService";
import { makeCrossingUrlFor } from "HOC/withApiService";
import Button from "components/Button";
import intl from "react-intl-universal";
import TextField from "components/TextField";
import Label from "components/Label";
import Icon from "components/Icon";
import Form from "components/Form";
import Columns from "components/Columns";
import Column from "components/Column";
import Select from "components/Select";
import PageContainer from "components/PageContainer";
import DatePicker from "components/DatePicker";
import ButtonGroup from "components/ButtonGroup";
import Switch from "components/Switch";
import { routePaths } from "containers/Routes";
import "./Search.css";

const passportTypeOptions = [
  {
    displayName: "normal",
    value: "normal"
  },
  {
    displayName: "diplomatic",
    value: "diplomatic"
  },
  {
    displayName: "special",
    value: "special"
  }
];
const { CITIZEN, PILGRIM, ALIEN, VISITOR } = travelerTypes;
const travelerTypeOptions = [
  {
    displayName: CITIZEN,
    value: CITIZEN
  },
  {
    displayName: ALIEN,
    value: ALIEN
  },
  {
    displayName: VISITOR,
    value: VISITOR
  },
  {
    displayName: PILGRIM,
    value: PILGRIM
  }
];
const { ARRIVAL } = crossingTypes;
const { SEARCH } = crossingActions;
const visaIssuanceOptions = [
  { displayName: "MOI", value: "MOI" },
  { displayName: "MOFA", value: "MOFA" }
];

class Search extends React.Component {
  formEl = React.createRef();
  _setState = this.setState;
  state = {
    currentState: "idle"
  };

  componentDidMount() {
    // this.getLookups();

    //todo: remove
    Array.from(this.formEl.current).forEach(elem => {
      if (elem instanceof HTMLInputElement) {
        const elemName = elem.name;
        if (elemName === "personId") {
          setInputValue(elem, "1119486650");
        }
        if (elemName === "idVersion") {
          setInputValue(elem, "1");
        }
      }
    });
    //////////////////////////////////////////////////////
  }

  getLookups = () => {
    // todo: get all lookups
  };

  selectTravelerType = ({ target }) => {
    this.props.crossingContext._setState({
      travelerType: target.value
    });
  };

  componentWillUnmount() {
    this._setState = null;
  }

  makeSearchTravelerData = () => {
    const { travelerType, crossingType } = this.props.crossingContext;

    let result = {};

    const elems = Array.from(this.formEl.current.elements);

    switch (travelerType) {
      case CITIZEN:
        result.searchCitizenCrossingInput = {
          crossingType: crossingType.toUpperCase()
        };
        elems.forEach(field => {
          const fieldName = field.name;
          const fieldValue = field.value;
          if (
            fieldName === "travelDate" ||
            fieldName === "personId" || //1119486650
            fieldName === "idVersion" //1
          )
            result.searchCitizenCrossingInput[fieldName] = fieldValue;
        });
        return result;

      case VISITOR:
        return {};

      case PILGRIM:
        return {};

      case ALIEN:
        return {
          searchCitizenCrossingInput: {
            personId: 1119486650,
            idVersion: 0,
            crossingType: "ARRIVAL",
            travelDate: "2018-10-03T17:00:00.000-0700"
          }
        };

      default:
        return {};
    }
  };

  searchCrossing = async () => {
    const { crossingContext } = this.props;

    const searchData = this.makeSearchTravelerData();

    const url = makeCrossingUrlFor(
      SEARCH,
      crossingContext.travelerType,
      crossingContext.crossingType
    );
    await crossingContext.searchCrossing({ data: searchData, url });

    // const searchRes = await sessionContext.post({
    //   url: makeCrossingUrlFor(
    //     SEARCH,
    //     crossingContext.travelerType,
    //     crossingContext.crossingType
    //   ),
    //   data: searchData
    // });

    // if (searchRes.xbcErrors) {
    //   searchRes.xbcErrors.forEach(error => {
    //     sessionContext.addNotification({ message: error.desc });
    //   });
    // }

    // if (searchRes.citizenDetailsOutput) {
    //   crossingContext._setState({
    //     ...searchRes.citizenDetailsOutput,
    //     travelDate: searchData.searchCitizenCrossingInput.travelDate
    //   });
    // }
  };

  searchMovements = async () => {
    const { crossingContext, history } = this.props;

    const movements = await crossingContext.searchMovements();

    if (movements.length === 1) {
      history.push(routePaths.crossing.searchResult);
    } else {
      history.push(routePaths.crossing.tripInfo);
    }
  };

  getTravelerId = async () => {
    const {
      apiService: { post, urls },
      crossingContext,
      history
    } = this.props;
    try {
      // const res = await post({
      //   url: "url to get traveler id",
      //   data: {
      //     visaNumber: Array.from(this.formEl.current.elements).filter(
      //       el => el.fieldName === "visaNumber"
      //     )[0].value
      //   }
      // });
      //set traveler id to response
      Array.from(this.formEl.current.elements).some(el => {
        el.name === "personId" && setInputValue(el, "12345");
      });
    } catch (e) {
      console.error("------failed to get traveler id:", e);
    }
  };

  next = () => {
    this._setState &&
      this._setState(
        {
          currentState: "loading"
        },
        async () => {
          const { sessionContext, history } = this.props;
          try {
            await this.searchCrossing();
            await this.searchMovements();
          } catch (e) {
            console.log("---- failed to search:", e);
            sessionContext.addNotification({ message: e.message || e });
          } finally {
            this._setState &&
              this._setState({
                currentState: "idle"
              });
          }
        }
      );
  };

  clearFields = () => {
    Array.from(this.formEl.current.el).forEach(elem => {
      if (elem instanceof HTMLInputElement) {
        elem.name !== "travelDate" && setInputValue(elem, "");
      }
    });
  };

  addVisitor = async () => {
    console.log("----------adding visitor");
  };

  render() {
    const { crossingType, travelerType } = this.props.crossingContext;
    const { hasRoles } = this.props.sessionContext;
    const { currentState } = this.state;
    const title =
      crossingType === ARRIVAL
        ? intl.get("entrySearchTitle")
        : intl.get("exitSearchTitle");
    return (
      <PageContainer title={title}>
        <div id="search" className="container">
          <Form forwardRef={this.formEl} style={{ margin: "0.75rem" }}>
            <Columns>
              <Column className="is-narrow">
                <DatePicker
                  id="travelDate"
                  isOutsideRange={
                    hasRoles("RBC.BC01.CHANGE.XDATE")
                      ? day =>
                          day.isAfter(moment()) ||
                          day.isBefore(moment().subtract(30, "days"))
                      : function() {}
                  }
                  label={<Label>{intl.get("travelDate")}</Label>}
                />
              </Column>
              <Column style={{ display: "flex", alignItems: "flex-end" }}>
                {travelerTypeOptions.map(o => (
                  <Button
                    name="travelerType"
                    value={o.value}
                    onClick={this.selectTravelerType}
                    key={o.value}
                    className={travelerType === o.value ? "is-primary" : ""}
                  >
                    {intl.get(o.displayName)}
                  </Button>
                ))}
              </Column>
            </Columns>
            {(travelerType === VISITOR || travelerType === PILGRIM) && (
              <React.Fragment>
                <Columns>
                  <Column className="is-narrow">
                    <DatePicker
                      disabled={currentState === "loading"}
                      label={<Label>{intl.get("birthDate")}</Label>}
                    />
                  </Column>
                  <Column className="is-narrow">
                    <div style={{ display: "flex", alignItems: "flex-end" }}>
                      <TextField
                        disabled={currentState === "loading"}
                        name="visaNumber"
                        label={<Label>{intl.get("visaNumber")}</Label>}
                      />
                      <a
                        style={{ marginBottom: "0.75rem" }}
                        onClick={this.getTravelerId}
                      >
                        <Icon className="is-medium" icon="fa fa-search" />
                      </a>
                    </div>
                  </Column>
                  {travelerType === VISITOR && (
                    <Column className="is-narrow">
                      <Select
                        options={visaIssuanceOptions.map(o => ({
                          ...o,
                          displayName: intl.get(o.displayName).d(o.displayName)
                        }))}
                        disabled={currentState === "loading"}
                        label={<Label>{intl.get("visaIssuance")}</Label>}
                      />
                    </Column>
                  )}
                </Columns>
              </React.Fragment>
            )}
            <Columns>
              <Column className="is-narrow">
                <TextField
                  disabled={currentState === "loading"}
                  type="number"
                  name="personId"
                  label={<Label>{intl.get("travelerId")}</Label>}
                />
              </Column>
              {travelerType !== VISITOR &&
                travelerType !== PILGRIM && (
                  <Column className="is-narrow">
                    <TextField
                      disabled={currentState === "loading"}
                      name="idVersion"
                      label={<Label>{intl.get("travelerIdVersion")}</Label>}
                    />
                  </Column>
                )}
            </Columns>
            {travelerType === CITIZEN && (
              <Columns>
                <Column className="padding-top-none is-narrow">
                  <TextField
                    disabled={currentState === "loading"}
                    name="passportNumber"
                    label={<Label>{intl.get("passportNumber")}</Label>}
                  />
                </Column>
                <Column className="padding-top-none is-narrow">
                  <Select
                    options={passportTypeOptions.map(o => ({
                      ...o,
                      displayName: intl.get(o.displayName).d(o.displayName)
                    }))}
                    name="passportType"
                    disabled={currentState === "loading"}
                    defaultValue={intl.get("chooseDocumentType")}
                    label={<Label>{intl.get("passportType")}</Label>}
                  />
                </Column>
              </Columns>
            )}
            {(travelerType === ALIEN ||
              travelerType === VISITOR ||
              travelerType === PILGRIM) && (
              <React.Fragment>
                <Columns>
                  <Column className="is-narrow">
                    <TextField
                      disabled={currentState === "loading"}
                      label={<Label>{intl.get("documentNumber")}</Label>}
                    />
                  </Column>
                  {travelerType === VISITOR &&
                    hasRoles("RBC.BC01.AUTH.GEN.DOCO.NUM") && (
                      <Column className="is-narrow">
                        <Switch
                          label={<Label>{intl.get("withoutDocument")}</Label>}
                        />
                      </Column>
                    )}
                </Columns>
                <Columns>
                  <Column className="is-narrow">
                    <TextField
                      label={<Label>{intl.get("nationalityCode")}</Label>}
                    />
                  </Column>
                  <Column className="is-narrow">
                    <Select
                      options={passportTypeOptions.map(o => ({
                        ...o,
                        displayName: intl.get(o.displayName).d(o.displayName)
                      }))}
                      disabled={currentState === "loading"}
                      defaultValue={intl.get("chooseDocumentType")}
                      label={<Label>{intl.get("nationality")}</Label>}
                    />
                  </Column>
                </Columns>
              </React.Fragment>
            )}
            <div
              style={{
                display: "flex",
                justifyContent:
                  crossingType === ARRIVAL &&
                  (travelerType === VISITOR || travelerType === PILGRIM)
                    ? "space-between"
                    : "flex-end",
                marginTop: "2rem"
              }}
            >
              {crossingType === ARRIVAL &&
                (travelerType === VISITOR || travelerType === PILGRIM) && (
                  <Link
                    className="button has-shadow"
                    to={routePaths.crossing.specialSearch}
                  >
                    <Icon icon="fa fa-search" />
                    &nbsp;
                    {intl.get("specialSearch")}
                  </Link>
                )}
              <div className="buttons" style={{ justifyContent: "flex-end" }}>
                <Link className="button has-shadow" to={routePaths.home}>
                  {intl.get("cancelBtn").d("Cancel")}
                </Link>
                {crossingType === ARRIVAL &&
                  (travelerType === VISITOR || travelerType === PILGRIM) && (
                    <React.Fragment>
                      <Button
                        className={
                          "is-primary has-shadow" +
                          (currentState === "loading" ? " is-loading" : "")
                        }
                        onClick={this.addVisitor}
                      >
                        <Icon icon="fa fa-plus" />
                        &nbsp;
                        {intl.get("addVisitor").d("Add Visitor")}
                      </Button>
                    </React.Fragment>
                  )}
                <Button
                  className={
                    "is-primary has-shadow" +
                    (currentState === "loading" ? " is-loading" : "")
                  }
                  disabled={!travelerType}
                  dataTest="search-traveler-btn"
                  onClick={this.next}
                >
                  {intl.get("searchBtn").d("Search")}
                </Button>
              </div>
            </div>
          </Form>
        </div>
      </PageContainer>
    );
  }
}

export { Search };

export default compose(
  // withApiService,
  withCrossingContext,
  withSessionContext
)(Search);
