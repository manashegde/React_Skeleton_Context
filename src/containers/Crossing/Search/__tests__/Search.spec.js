import React from "react";
import PropTypes from "prop-types";
import { shallow, mount } from "enzyme";
import createRouterContext from "react-router-test-context";
import toJson from "enzyme-to-json";
import { Search } from "containers/Crossing/Search";
import { urls } from "HOC/withApiService";
import { resolve } from "path";

Search.contextTypes = {
  router: PropTypes.object
};

const props = {
  location: {
    state: { crossingType: "entry" }
  },
  history: { push: jest.fn() },
  apiService: {
    urls,
    post: jest.fn()
  },
  sessionContext: {
    addNotification: jest.fn()
  },
  crossingContext: {
    _setState: jest.fn()
  }
};

const setup = () => {
  const context = createRouterContext();
  const wrapper = mount(<Search {...props} />, { context });
  const instance = wrapper.instance();
  return { wrapper, instance };
};

afterEach(() => {
  props.apiService.post = jest.fn();
});

describe("Home", () => {
  it("should match snapshot", () => {
    const { wrapper } = setup();
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });

  it("should call api to search for citizen and go to the next page", async () => {
    props.apiService.post = jest
      .fn()
      .mockReturnValueOnce(Promise.resolve(true))
      .mockReturnValueOnce(Promise.resolve(true));
    const { wrapper, instance } = setup();

    await instance.doSearchCitizen();
    await wrapper.update();
    expect(props.apiService.post).toHaveBeenCalled();
    expect(props.apiService.post.mock.calls.length).toBe(2); //first to search citizen. second to get flight info
    expect(props.history.push).toHaveBeenCalled();
    expect(props.crossingContext._setState).toHaveBeenCalled();
    expect(props.sessionContext.addNotification).not.toHaveBeenCalled();
  });
  it("should show notification on search api failure", async () => {
    props.apiService.post = jest
      .fn()
      .mockReturnValueOnce(Promise.reject(true))
      .mockReturnValueOnce(Promise.resolve(true));
    const { wrapper, instance } = setup();

    await instance.doSearchCitizen();
    wrapper.update();
    expect(props.sessionContext.addNotification).toHaveBeenCalled();
  });
});
