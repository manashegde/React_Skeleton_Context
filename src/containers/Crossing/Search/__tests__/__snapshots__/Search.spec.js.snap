// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
<Search
  apiService={
    Object {
      "post": [Function],
      "urls": Object {
        "completeCrossing": "/api/citizen/completeCrossing",
        "fetchConsolidatedHitReport": "/api/watchlist/fetchConsolidatedHitReport",
        "fetchExpectedFlights": "/api/flight/fetchExpected",
        "login": "/api/login",
        "searchCitizen": "/api/citizen/searchCitizen",
        "setActualFlight": "/api/flight/setActual",
      },
    }
  }
  crossingContext={
    Object {
      "_setState": [Function],
    }
  }
  history={
    Object {
      "push": [Function],
    }
  }
  location={
    Object {
      "state": Object {
        "crossingType": "entry",
      },
    }
  }
  sessionContext={
    Object {
      "addNotification": [Function],
    }
  }
>
  <Box
    className="form-box"
  >
    <div
      className="box form-box"
    >
      <SubTitle
        dataTest="search-form-title"
      >
        <h5
          className="subtitle is-5"
          data-test="search-form-title"
        />
      </SubTitle>
      <br />
      <br />
      <a
        onClick={[Function]}
        style={
          Object {
            "color": "white",
          }
        }
      />
      <br />
      <br />
      <Form>
        <form
          noValidate={true}
          onSubmit={[Function]}
        >
          <Columns>
            <div
              className="columns"
            >
              <Column>
                <div
                  className="column"
                >
                  <TextField
                    disabled={false}
                    label={<Label />}
                    name="travelDate"
                  >
                    <div
                      className="field"
                    >
                      <Label>
                        <label
                          className="label"
                        />
                      </Label>
                      <div
                        className="control"
                      >
                        <input
                          className="input"
                          disabled={false}
                          name="travelDate"
                          onChange={[Function]}
                          type="text"
                          value=""
                        />
                        <p
                          className="help is-warning"
                          style={
                            Object {
                              "fontSize": "15px",
                            }
                          }
                        />
                      </div>
                    </div>
                  </TextField>
                </div>
              </Column>
              <Column>
                <div
                  className="column"
                >
                  <Select
                    defaultValue=""
                    disabled={false}
                    label={<Label />}
                    name="travelerType"
                    onChange={[Function]}
                    options={
                      Array [
                        Object {
                          "displayName": "citizen",
                          "value": "citizen",
                        },
                        Object {
                          "displayName": "alien",
                          "value": "alien",
                        },
                        Object {
                          "displayName": "visitor",
                          "value": "visitor",
                        },
                        Object {
                          "displayName": "pilgrim",
                          "value": "pilgrim",
                        },
                      ]
                    }
                  >
                    <div
                      className="field"
                    >
                      <Label>
                        <label
                          className="label"
                        />
                      </Label>
                      <div
                        className="control"
                      >
                        <div
                          className="select"
                        >
                          <select
                            disabled={false}
                            name="travelerType"
                            onChange={[Function]}
                          >
                            <option
                              key="citizen"
                              value="citizen"
                            >
                              citizen
                            </option>
                            <option
                              key="alien"
                              value="alien"
                            >
                              alien
                            </option>
                            <option
                              key="visitor"
                              value="visitor"
                            >
                              visitor
                            </option>
                            <option
                              key="pilgrim"
                              value="pilgrim"
                            >
                              pilgrim
                            </option>
                          </select>
                        </div>
                      </div>
                    </div>
                  </Select>
                </div>
              </Column>
            </div>
          </Columns>
          <Columns>
            <div
              className="columns"
            >
              <Column>
                <div
                  className="column"
                >
                  <TextField
                    disabled={false}
                    label={<Label />}
                  >
                    <div
                      className="field"
                    >
                      <Label>
                        <label
                          className="label"
                        />
                      </Label>
                      <div
                        className="control"
                      >
                        <input
                          className="input"
                          disabled={false}
                          onChange={[Function]}
                          type="text"
                          value=""
                        />
                        <p
                          className="help is-warning"
                          style={
                            Object {
                              "fontSize": "15px",
                            }
                          }
                        />
                      </div>
                    </div>
                  </TextField>
                </div>
              </Column>
              <Column>
                <div
                  className="column"
                >
                  <TextField
                    disabled={false}
                    label={<Label />}
                  >
                    <div
                      className="field"
                    >
                      <Label>
                        <label
                          className="label"
                        />
                      </Label>
                      <div
                        className="control"
                      >
                        <input
                          className="input"
                          disabled={false}
                          onChange={[Function]}
                          type="text"
                          value=""
                        />
                        <p
                          className="help is-warning"
                          style={
                            Object {
                              "fontSize": "15px",
                            }
                          }
                        />
                      </div>
                    </div>
                  </TextField>
                </div>
              </Column>
            </div>
          </Columns>
          <br />
          <br />
          <br />
          <div
            className="buttons"
            style={
              Object {
                "justifyContent": "space-between",
              }
            }
          >
            <Link
              className="button"
              to="/authenticated/home"
            >
              <Link
                className="button"
                replace={false}
                to="/authenticated/home"
              >
                <a
                  className="button"
                  href="/authenticated/home"
                  onClick={[Function]}
                >
                  Cancel
                </a>
              </Link>
            </Link>
            <Button
              className="is-primary"
              dataTest="search-traveler-btn"
              onClick={[Function]}
            >
              <button
                className="button is-primary"
                data-test="search-traveler-btn"
                onClick={[Function]}
              >
                Search
              </button>
            </Button>
          </div>
        </form>
      </Form>
    </div>
  </Box>
</Search>
`;
